modules/hdf5_handler/HDF5VlenAtomicArray.cc

```diff
@@ -0,0 +1,262 @@
+// This file is part of the hdf5_handler implementing for the default option
+// Copyright (c) The HDF Group, Inc. and OPeNDAP, Inc.
+//
+// This is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License as published by the Free
+// Software Foundation; either version 2.1 of the License, or (at your
+// option) any later version.
+//
+// This software is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+// License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+//
+// You can contact OPeNDAP, Inc. at PO Box 112, Saunderstown, RI. 02874-0112.
+// You can contact The HDF Group, Inc. at 410 E University Ave,
+// Suite 200, Champaign, IL 61820  
+
+/////////////////////////////////////////////////////////////////////////////
+/// \file HDF5VlenAtomicArray.cc
+/// \brief The implementation of the retrieval of data values for Vlen integer and float array data.
+/// Currently this only applies to ACOS level 2 and OCO2 level 1B data.
+/// \author Kent Yang <myang6@hdfgroup.org>
+///
+/// Copyright (C) The HDF Group
+///
+/// All rights reserved.
+
+#include <iostream>
+#include <memory>
+#include <cassert>
+#include <BESDebug.h>
+#include <libdap/InternalErr.h>
+
+#include "HDF5RequestHandler.h"
+#include "HDF5VlenAtomicArray.h"
+
+using namespace std;
+using namespace libdap;
+
+BaseType *HDF5VlenAtomicArray::ptr_duplicate()
+{
+    auto HDF5VlenAtomicArray_unique = make_unique<HDF5VlenAtomicArray>(*this);
+    return HDF5VlenAtomicArray_unique.release();
+}
+
+HDF5VlenAtomicArray::HDF5VlenAtomicArray(const string & n, const string &d, BaseType * v, bool vlen_index) :
+    HDF5Array(n, d, v),is_vlen_index(vlen_index) {
+}
+
+bool HDF5VlenAtomicArray::read()
+{
+    BESDEBUG("h5","Coming to HDF5VlenAtomicArray read "<<endl);
+
+    read_vlen_internal(is_vlen_index);
+    
+    return true;
+}
+
+void HDF5VlenAtomicArray::read_vlen_internal(bool vlen_index) {
```

Comment:
If it's complicated to track which of the HDF5 objects (e.g., vlen_memtype) need to be closed when exceptions are thrown, maybe breaking this method up into parts would help?

Reply:
Will see. But mostly I will just add those H5?close() in the exception blocks.